README
NAME
    Text::Levenshtein::XS - Calculate edit distance based on insertion,
    deletion, substitution, and transposition

VERSION
    version 0.47

SYNOPSIS
        use Text::Levenshtein::XS qw/distance/;

        print distance('Neil','Niel');
        # prints 2

DESCRIPTION
    Returns the number of edits (insert,delete,substitute) required to turn
    the source string into the target string. XS implementation (requires a
    C compiler). Works correctly with utf8.

        use Text::Levenshtein::XS qw/distance/;
        use utf8;

        distance('ⓕⓞⓤⓡ','ⓕⓤⓞⓡ'), 
        # prints 2

METHODS
  distance
    Arguments: $source_text, $target_text, (optional) $max_distance
    Return Value: Int $edit_distance || undef (if max_distance is exceeded)

    Returns: int that represents the edit distance between the two argument,
    or undef if $max_distance threshold is exceeded.

    Takes the edit distance between a source and target string using XS 2
    vector implementation.

        use Text::Levenshtein::XS qw/distance/;
        print distance('Neil','Niel');
        # prints 2

    Stops calculations and returns undef if $max_distance is set, non-zero
    (0 = no limit), and the algorithm has determined the final distance will
    be greater than $max_distance.

        my $distance = distance('Neil','Niel',1);
        print (defined $distance) ? $distance : "Exceeded max distance";
        # prints "Exceeded max distance"

NOTES
    Drop in replacement for Text::LevenshteinXS

SEE ALSO
    *   Text::Levenshtein::Damerau

    *   Text::Levenshtein::Damerau::PP

    *   Text::Levenshtein::Damerau::XS

    *   Text::Fuzzy

    *   Text::Levenshtein::Flexible

REPOSITORY
    <https://github.com/ugexe/Text--Levenshtein--XS>

BUGS
    Please report bugs to:

    <https://github.com/ugexe/Text--Levenshtein--XS/issues>

AUTHOR
    ugexe <ugexe@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Nick Logan.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



CHANGES
Revision history for Perl module Text::Levenshtein::XS

{{$NEXT}}

0.44      2014-10-16 00:25:19-04:00 America/New_York
    - Was wrong about the brackets. Switching back to regular INIT as it seemed to pass on a later build :/
    - Update dist.ini TravisYML to support_builddir = 1
    - (branch~1 error on TravisCI) .travis.yml add: git checkout $TRAVIS_BRANCH && git reset --hard $TRAVIS_COMMIT (thanks haarg)
    
0.43      2014-10-15 10:45:00-04:00 America/New_York
    - brackets in XS init for C89 compatability

0.428     2014-10-14 22:44:32-04:00 America/New_York
    - Removed travis-ci/coveralls badges until I figure out Dist::Zilla better

0.427     2014-10-14 22:31:47-04:00 America/New_York
    - Update .travis.yml to build/test properly with Dist::Zilla

0.426     2014-10-14 15:19:12-04:00 America/New_York
    - Dist::Zilla
    - New max distance algorithm bug fix (github issue #7). Caught by nglenn@cpan.org

0.42 2014-10-12 14:07:31-04:00 America/New_York
    - CPAN release of 0.41_02 features/updates

0.41_02 2014-10-12 12:04:49-04:00 America/New_York
    - max_distance functionality added
    - tests for max_distance
    - Change how the return value is placed on the stack (now using XPUSH instead of modifying TARG)
    - memory leak author test

0.40 2014-10-11 01:20:21-04:00 America/New_York
    - I guess 0.3 > 0.31

0.31 2014-10-11 01:11:46-04:00 America/New_York
    - Fixed: https://rt.cpan.org/Public/Bug/Display.html?id=99413
    - Using 2 row algorithm instead of slength x tlength matrix

0.30_01 2014-10-10 15:03:02-04:00 America/New_York
    - https://rt.cpan.org/Public/Bug/Display.html?id=99413 Not sure the source of this bug. Just putting up to test.

0.30 2013-01-20 23:38:11-04:00 America/New_York
    - alloca changed to malloc
    - associated memory managment
    - ppm should build now

0.20  2013-01-18 23:22:31-04:00 America/New_York
    - MIN/MAX macros added
    - loops combined

0.10 2013-10-14 01:20:51-04:00 America/New_York
    - Initial release. 






.build/testing/Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.020.
use strict;
use warnings;

use 5.008;

use ExtUtils::MakeMaker ;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Calculate edit distance based on insertion, deletion, substitution, and transposition",
  "AUTHOR" => "ugexe <ugexe\@cpan.org>",
  "CONFIGURE_REQUIRES" => {},
  "DISTNAME" => "Text-Levenshtein-XS",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Text::Levenshtein::XS",
  "PREREQ_PM" => {},
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.470",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Test::More" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





.build/testing/lib/Text/Levenshtein/XS.pm
package Text::Levenshtein::XS;

# generated by Dist::Zilla::Plugin::OurPkgVersion
our $VERSION = '0.470'; # TRIAL VERSION
our $AUTHORITY = 'cpan:UGEXE'; # AUTHORITY

use 5.008;
use strict;
use warnings FATAL => 'all';
require Exporter;
require XSLoader;
XSLoader::load('Text::Levenshtein::XS', $Text::Levenshtein::XS::VERSION);

@Text::Levenshtein::XS::ISA       = qw/Exporter/;
@Text::Levenshtein::XS::EXPORT_OK = qw/distance/;



sub distance {
    return Text::Levenshtein::XS::xs_distance( [unpack('U*', shift)], [unpack('U*', shift)], shift || 0);
}



1;

=pod

=encoding UTF-8

=head1 NAME

Text::Levenshtein::XS - Calculate edit distance based on insertion, deletion, substitution, and transposition

=head1 VERSION

version 0.470

=head1 SYNOPSIS

    use Text::Levenshtein::XS qw/distance/;

    print distance('Neil','Niel');
    # prints 2

=head1 DESCRIPTION

Returns the number of edits (insert,delete,substitute) required to turn the source string into the target string. XS implementation (requires a C compiler). Works correctly with utf8.

    use Text::Levenshtein::XS qw/distance/;
    use utf8;

    distance('ⓕⓞⓤⓡ','ⓕⓤⓞⓡ'), 
    # prints 2

=for Pod::Coverage dl_load_flags xs_distance

=head1 METHODS

=head2 distance

=over 4

=item Arguments: $source_text, $target_text, (optional) $max_distance

=item Return Value: Int $edit_distance || undef (if max_distance is exceeded)

=back

Returns: int that represents the edit distance between the two argument, or undef if $max_distance threshold is exceeded.

Takes the edit distance between a source and target string using XS 2 vector implementation.

    use Text::Levenshtein::XS qw/distance/;
    print distance('Neil','Niel');
    # prints 2

Stops calculations and returns undef if $max_distance is set, non-zero (0 = no limit), and the algorithm has determined the final distance will be greater than $max_distance.

    my $distance = distance('Neil','Niel',1);
    print (defined $distance) ? $distance : "Exceeded max distance";
    # prints "Exceeded max distance"

=head1 NOTES

Drop in replacement for L<Text::LevenshteinXS>

=head1 SEE ALSO

=over 4

=item * L<Text::Levenshtein::Damerau>

=item * L<Text::Levenshtein::Damerau::PP>

=item * L<Text::Levenshtein::Damerau::XS>

=item * L<Text::Fuzzy>

=item * L<Text::Levenshtein::Flexible>

=back

=head1 REPOSITORY

L<https://github.com/ugexe/Text--Levenshtein--XS>

=head1 BUGS

Please report bugs to:

L<https://github.com/ugexe/Text--Levenshtein--XS/issues>

=head1 AUTHOR

ugexe <ugexe@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Nick Logan.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut

__END__



# ABSTRACT: Calculate edit distance based on insertion, deletion, substitution, and transposition



.build/testing/examples/fuzzy_name_search.pl
use strict;
use warnings;
use Text::Levenshtein::XS qw/distance/;

my @names = ( 
    'Angela Smarts',
    'Angela Sharron',
    'Andrew North',
    'Andy North',
    'Andy Norths',
    'Ameila Anderson',
);

print "[NAMES]: " . join(', ',@names) . "\n\n";
print "Enter a name to fuzzy search against: ";
my $fuzzy_name = <>;
chomp($fuzzy_name);
print "\n";

my $best_match = "";
my $best_distance;

foreach my $name (@names) {
    my $distance = distance($fuzzy_name,$name);
    print "*$name - $distance\n";

    if( !defined $best_distance || $distance < $best_distance ) {
        $best_match = $name;
        $best_distance = $distance;
    }
}

print "\n\nDamerau-Levenshtein search result: " . $best_match . " with a distance of " . $best_distance . "\n";

1;

lib/Text/Levenshtein/XS.pm
package Text::Levenshtein::XS;

# generated by Dist::Zilla::Plugin::OurPkgVersion
# VERSION
# AUTHORITY

use 5.008;
use strict;
use warnings FATAL => 'all';
require Exporter;
require XSLoader;
XSLoader::load('Text::Levenshtein::XS', $Text::Levenshtein::XS::VERSION);

@Text::Levenshtein::XS::ISA       = qw/Exporter/;
@Text::Levenshtein::XS::EXPORT_OK = qw/distance/;



sub distance {
    return Text::Levenshtein::XS::xs_distance( [unpack('U*', shift)], [unpack('U*', shift)], shift || 0);
}



1;



__END__



# ABSTRACT: Calculate edit distance based on insertion, deletion, substitution, and transposition

=for Pod::Coverage dl_load_flags xs_distance

=encoding utf8

=head1 SYNOPSIS

    use Text::Levenshtein::XS qw/distance/;

    print distance('Neil','Niel');
    # prints 2

=head1 DESCRIPTION

Returns the number of edits (insert,delete,substitute) required to turn the source string into the target string. XS implementation (requires a C compiler). Works correctly with utf8.

    use Text::Levenshtein::XS qw/distance/;
    use utf8;

    distance('ⓕⓞⓤⓡ','ⓕⓤⓞⓡ'), 
    # prints 2

=head1 METHODS

=head2 distance

=over 4

=item Arguments: $source_text, $target_text, (optional) $max_distance

=item Return Value: Int $edit_distance || undef (if max_distance is exceeded)

=back

Returns: int that represents the edit distance between the two argument, or undef if $max_distance threshold is exceeded.

Takes the edit distance between a source and target string using XS 2 vector implementation.

    use Text::Levenshtein::XS qw/distance/;
    print distance('Neil','Niel');
    # prints 2

Stops calculations and returns undef if $max_distance is set, non-zero (0 = no limit), and the algorithm has determined the final distance will be greater than $max_distance.

    my $distance = distance('Neil','Niel',1);
    print (defined $distance) ? $distance : "Exceeded max distance";
    # prints "Exceeded max distance"

=head1 NOTES

Drop in replacement for L<Text::LevenshteinXS>

=head1 SEE ALSO

=over 4

=item * L<Text::Levenshtein::Damerau>

=item * L<Text::Levenshtein::Damerau::PP>

=item * L<Text::Levenshtein::Damerau::XS>

=item * L<Text::Fuzzy>

=item * L<Text::Levenshtein::Flexible>

=back

=head1 REPOSITORY

L<https://github.com/ugexe/Text--Levenshtein--XS>

=head1 BUGS

Please report bugs to:

L<https://github.com/ugexe/Text--Levenshtein--XS/issues>

=cut


examples/fuzzy_name_search.pl
#
# This file is part of Text-Levenshtein-XS
#
# This software is copyright (c) 2014 by Nick Logan.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
use strict;
use warnings;
use Text::Levenshtein::XS qw/distance/;

my @names = ( 
    'Angela Smarts',
    'Angela Sharron',
    'Andrew North',
    'Andy North',
    'Andy Norths',
    'Ameila Anderson',
);

print "[NAMES]: " . join(', ',@names) . "\n\n";
print "Enter a name to fuzzy search against: ";
my $fuzzy_name = <>;
chomp($fuzzy_name);
print "\n";

my $best_match = "";
my $best_distance;

foreach my $name (@names) {
    my $distance = distance($fuzzy_name,$name);
    print "*$name - $distance\n";

    if( !defined $best_distance || $distance < $best_distance ) {
        $best_match = $name;
        $best_distance = $distance;
    }
}

print "\n\nDamerau-Levenshtein search result: " . $best_match . " with a distance of " . $best_distance . "\n";

1;

Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.020.
use strict;
use warnings;

use 5.008;

use ExtUtils::MakeMaker ;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Calculate edit distance based on insertion, deletion, substitution, and transposition",
  "AUTHOR" => "ugexe <ugexe\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Text-Levenshtein-XS",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Text::Levenshtein::XS",
  "PREREQ_PM" => {
    "Exporter" => 0,
    "XSLoader" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => 0,
    "utf8" => 0
  },
  "VERSION" => "0.47",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Test::More" => 0,
  "XSLoader" => 0,
  "strict" => 0,
  "utf8" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



